src/wire/bulk_report.rs

```diff
@@ -0,0 +1,349 @@
+    /// with a payload length of `n_log_bytes` + `n_extension_bytes`
+    pub fn buffer_len(n_log_bytes: usize, n_extension_bytes: usize) -> usize {
+        field::PAYLOAD.start + n_log_bytes + n_extension_bytes
+    }
+
+    /// Return the length of the bulk report payload
+    pub fn payload_len(&self) -> usize {
+        let n_log_bytes = self.n_log_bytes() as usize;
+        let n_extension_bytes = self.n_extension_bytes() as usize;
+        n_log_bytes + n_extension_bytes
+    }
+
+    /// Return the `fingerprint` field
+    #[inline]
+    pub fn fingerprint(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::FINGERPRINT])
+    }
+
+    /// Return the `probe_id` field
+    #[inline]
+    pub fn probe_id(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::PROBE_ID])
+    }
+
+    /// Return the `n_log_bytes` field
+    #[inline]
+    pub fn n_log_bytes(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::N_LOG_BYTES])
+    }
+
+    /// Return the `n_extension_bytes` field
+    #[inline]
+    pub fn n_extension_bytes(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        LittleEndian::read_u32(&data[field::N_EXT_BYTES])
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> BulkReport<&'a T> {
+    /// Return a pointer to the payload
+    #[inline]
+    pub fn payload(&self) -> &'a [u8] {
+        let data = self.buffer.as_ref();
+        &data[field::PAYLOAD]
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> BulkReport<T> {
+    /// Set the `fingerprint` field
+    #[inline]
+    pub fn set_fingerprint(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u32(&mut data[field::FINGERPRINT], value);
+    }
+
+    /// Set the `probe_id` field
+    #[inline]
+    pub fn set_probe_id(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u32(&mut data[field::PROBE_ID], value);
+    }
+
+    /// Set the `n_log_bytes` field
+    #[inline]
+    pub fn set_n_log_bytes(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u32(&mut data[field::N_LOG_BYTES], value);
+    }
+
+    /// Set the `n_extension_bytes` field
+    #[inline]
+    pub fn set_n_extension_bytes(&mut self, value: u32) {
+        let data = self.buffer.as_mut();
+        LittleEndian::write_u32(&mut data[field::N_EXT_BYTES], value);
+    }
+
+    /// Return a mutable pointer to the payload
+    #[inline]
+    pub fn payload_mut(&mut self) -> &mut [u8] {
+        let data = self.buffer.as_mut();
+        &mut data[field::PAYLOAD]
+    }
+}
+
+impl<T: AsRef<[u8]>> AsRef<[u8]> for BulkReport<T> {
+    fn as_ref(&self) -> &[u8] {
+        self.buffer.as_ref()
+    }
+}
+
+impl<T: AsRef<[u8]> + AsMut<[u8]>> AsMut<[u8]> for BulkReport<T> {
+    fn as_mut(&mut self) -> &mut [u8] {
```

Comment:
I don't recall - why do we want to expose the raw bytes mutably? For alternate mutable overlays? Some other reason?